File: server/src/controllers/queueController.js
const supabase            = require('../db/supabase');
const { getQueueState, invalidateCache, getNextInQueue, calcWaitTime } = require('../services/queueService');
const { addClient, removeClient, broadcastQueue } = require('../services/sseService');
const { notifyTurnNear }  = require('../services/twilioService');

// ── SSE Endpoint ─────────────────────────────────────────────
async function liveQueue(req, res) {
  const { businessId } = req.params;

  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.setHeader('Access-Control-Allow-Origin', process.env.CLIENT_URL);
  res.flushHeaders();

  const clientId = `${Date.now()}-${Math.random()}`;
  addClient(businessId, clientId, res);

  // Send current state immediately on connect
  try {
    // Get active slot for this business today
    const today = new Date().toISOString().split('T')[0];
    const { data: slots } = await supabase
      .from('slots')
      .select('id')
      .eq('business_id', businessId)
      .eq('date', today)
      .eq('is_active', true);

    if (slots && slots.length > 0) {
      const slotId = slots[0].id;
      const queue = await getQueueState(slotId);
      res.write(`data: ${JSON.stringify({ slotId, queue })}\n\n`);
    } else {
      res.write(`data: ${JSON.stringify({ slotId: null, queue: [] })}\n\n`);
    }
  } catch { /* ignore initial load errors */ }

  req.on('close', () => removeClient(businessId, clientId));
}

// ── Call Next Token ──────────────────────────────────────────
async function callNext(req, res) {
  const { slotId } = req.body;

  // Mark currently serving as done
  await supabase
    .from('bookings')
    .update({ status: 'done' })
    .eq('slot_id', slotId)
    .eq('status', 'serving');

  await supabase.from('queue_events').insert({
    booking_id: null,
    event_type: 'done',
  });

  const next = await getNextInQueue(slotId);
  if (!next) {
    return res.json({ message: 'Queue is empty', next: null });
  }

  // Mark next as serving
  await supabase
    .from('bookings')
    .update({ status: 'serving' })
    .eq('id', next.id);

  await supabase.from('queue_events').insert({
    booking_id: next.id,
    event_type: 'serving',
  });

  // Check if anyone is 2 tokens away and notify them
  const queue = await getQueueState(slotId);
  const twoAway = queue.find(b => b.token_number === next.token_number + 2);
  if (twoAway) {
    // Get their phone
    const { data: user } = await supabase
      .from('users')
      .select('phone, name')
      .eq('id', twoAway.user_id)
      .single();
    if (user?.phone) {
      const { data: slot } = await supabase
        .from('slots')
        .select('businesses(name)')
        .eq('id', slotId)
        .single();
      await notifyTurnNear(user.phone, twoAway.token_number, slot?.businesses?.name || 'the service');
    }
  }

  await invalidateCache(slotId);
  const updatedQueue = await getQueueState(slotId);

  // Get business_id to broadcast
  const { data: slot } = await supabase
    .from('slots')
    .select('business_id')
    .eq('id', slotId)
    .single();

  broadcastQueue(slot.business_id, { slotId, queue: updatedQueue });
  res.json({ success: true, serving: next, queue: updatedQueue });
}

// ── Skip Token ───────────────────────────────────────────────
async function skipToken(req, res) {
  const { bookingId, slotId } = req.body;

  const { data: booking } = await supabase
    .from('bookings')
    .update({ status: 'skipped' })
    .eq('id', bookingId)
    .select()
    .single();

  await supabase.from('queue_events').insert({ booking_id: bookingId, event_type: 'skipped' });

  await invalidateCache(slotId);
  const queue = await getQueueState(slotId);

  const { data: slot } = await supabase
    .from('slots')
    .select('business_id')
    .eq('id', slotId)
    .single();

  broadcastQueue(slot.business_id, { slotId, queue });
  res.json({ success: true, booking, queue });
}

// ── Get Queue Status for a slot ──────────────────────────────
async function getQueueStatus(req, res) {
  const { slotId } = req.params;
  const queue = await getQueueState(slotId);

  // Get avg service time
  const { data: slot } = await supabase
    .from('slots')
    .select('businesses(avg_service_time)')
    .eq('id', slotId)
    .single();

  const avgTime = slot?.businesses?.avg_service_time || 10;

  const enriched = queue.map((b, i) => ({
    ...b,
    position: i + 1,
    estimated_wait: calcWaitTime(i, avgTime),
  }));

  res.json(enriched);
}

module.exports = { liveQueue, callNext, skipToken, getQueueStatus };

File: server/src/main/java/com/smartqueue/server/controller/QueueController.java
package com.smartqueue.server.controller;

import com.smartqueue.server.entity.Booking;
import com.smartqueue.server.entity.QueueEvent;
import com.smartqueue.server.entity.Slot;
import com.smartqueue.server.entity.User;
import com.smartqueue.server.repository.BookingRepository;
import com.smartqueue.server.repository.QueueEventRepository;
import com.smartqueue.server.repository.SlotRepository;
import com.smartqueue.server.repository.UserRepository;
import com.smartqueue.server.security.UserPrincipal;
import com.smartqueue.server.service.QueueService;
import com.smartqueue.server.service.SseService;
import com.smartqueue.server.service.TwilioService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import java.time.LocalDate;
import java.util.*;

@RestController
@RequestMapping("/api/queue")
public class QueueController {

    private final QueueService queueService;
    private final SseService sseService;
    private final TwilioService twilioService;
    private final SlotRepository slotRepository;
    private final BookingRepository bookingRepository;
    private final QueueEventRepository queueEventRepository;
    private final UserRepository userRepository;

    public QueueController(QueueService queueService,
                           SseService sseService,
                           TwilioService twilioService,
                           SlotRepository slotRepository,
                           BookingRepository bookingRepository,
                           QueueEventRepository queueEventRepository,
                           UserRepository userRepository) {
        this.queueService = queueService;
        this.sseService = sseService;
        this.twilioService = twilioService;
        this.slotRepository = slotRepository;
        this.bookingRepository = bookingRepository;
        this.queueEventRepository = queueEventRepository;
        this.userRepository = userRepository;
    }

    @Data
    public static class CallNextRequest {
        @NotNull(message = "slotId is required")
        private UUID slotId;
    }

    @Data
    public static class SkipTokenRequest {
        @NotNull(message = "bookingId is required")
        private UUID bookingId;
        @NotNull(message = "slotId is required")
        private UUID slotId;
    }

    @Data
    public static class CompleteServiceRequest {
        @NotNull(message = "bookingId is required")
        private UUID bookingId;
        @NotNull(message = "slotId is required")
        private UUID slotId;
    }

    // SSE Endpoint — no auth needed (public queue board)
    @GetMapping("/live/{businessId}")
    public SseEmitter liveQueue(@PathVariable UUID businessId,
                                @RequestParam(required = false) UUID slotId) {
        // Use a high timeout (e.g., 24 hours) to keep SSE connection alive
        SseEmitter emitter = new SseEmitter(24 * 60 * 60 * 1000L);
        String clientId = UUID.randomUUID().toString();

        sseService.addClient(businessId, clientId, emitter, slotId);

        try {
            UUID targetSlotId = slotId;
            if (targetSlotId == null) {
                // Get active slot for this business today
                List<Slot> slots = slotRepository.findByBusinessIdAndDate(businessId, LocalDate.now());
                Optional<Slot> activeSlot = slots.stream().filter(Slot::getIsActive).findFirst();
                if (activeSlot.isPresent()) {
                    targetSlotId = activeSlot.get().getId();
                }
            }

            if (targetSlotId != null) {
                List<Booking> queue = queueService.getQueueState(targetSlotId);
                emitter.send(SseEmitter.event().data(Map.of("slotId", targetSlotId, "queue", queue)));
            } else {
                emitter.send(SseEmitter.event().data(Map.of("slotId", "", "queue", Collections.emptyList())));
            }
        } catch (Exception e) {
            // Client closed early, removeClient will be called via onError/onCompletion
        }

        return emitter;
    }

    // Call Next Token
    @PostMapping("/next")
    @PreAuthorize("hasAnyRole('admin', 'staff')")
    @Transactional
    public ResponseEntity<?> callNext(@AuthenticationPrincipal UserPrincipal principal,
                                      @Valid @RequestBody CallNextRequest body) {
        UUID slotId = body.getSlotId();

        Optional<Slot> slotOpt = slotRepository.findById(slotId);
        if (slotOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Slot not found"));
        }
        Slot slot = slotOpt.get();

        // Enforce ownership checks
        if ("staff".equals(principal.getRole())) {
            if (principal.getBusinessId() == null || !principal.getBusinessId().equals(slot.getBusinessId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("error", "Access denied: this slot does not belong to your business"));
            }
        } else if ("admin".equals(principal.getRole())) {
            if (slot.getBusiness() == null || !principal.getId().equals(slot.getBusiness().getOwnerId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("error", "Access denied: this slot does not belong to your business"));
            }
        }

        // Fetch current serving booking and mark as done
        List<Booking> bookings = bookingRepository.findBySlotIdOrderByTokenNumberAsc(slotId);
        for (Booking b : bookings) {
            if ("serving".equals(b.getStatus())) {
                b.setStatus("done");
                bookingRepository.save(b);
            }
        }

        // Insert done event
        QueueEvent doneEvent = QueueEvent.builder()
                .bookingId(null)
                .eventType("done")
                .build();
        queueEventRepository.save(doneEvent);

        // Fetch next pending/arrived
        Optional<Booking> nextOpt = queueService.getNextInQueue(slotId);
        if (nextOpt.isEmpty()) {
            return ResponseEntity.ok(Map.of("message", "Queue is empty", "next", null));
        }

        Booking next = nextOpt.get();
        next.setStatus("serving");
        next = bookingRepository.save(next);

        // Insert serving event
        QueueEvent servingEvent = QueueEvent.builder()
                .bookingId(next.getId())
                .eventType("serving")
                .build();
        queueEventRepository.save(servingEvent);

        // Notify customer 2 tokens away
        final int currentTokenNum = next.getTokenNumber();
        List<Booking> queue = queueService.getQueueState(slotId);
        Optional<Booking> twoAwayOpt = queue.stream()
                .filter(b -> b.getTokenNumber() == currentTokenNum + 2)
                .findFirst();

        if (twoAwayOpt.isPresent()) {
            Booking twoAway = twoAwayOpt.get();
            Optional<User> userOpt = userRepository.findById(twoAway.getUserId());
            if (userOpt.isPresent() && userOpt.get().getPhone() != null && !userOpt.get().getPhone().isBlank()) {
                String businessName = slot.getBusiness() != null ?
                        slot.getBusiness().getName() : "the service";
                twilioService.notifyTurnNear(userOpt.get().getPhone(), twoAway.getTokenNumber(), businessName);
            }
        }

        queueService.invalidateCache(slotId);
        List<Booking> updatedQueue = queueService.getQueueState(slotId);

        // Broadcast update
        sseService.broadcastQueue(slot.getBusinessId(), slotId, Map.of("slotId", slotId, "queue", updatedQueue));

        return ResponseEntity.ok(Map.of("success", true, "serving", next, "queue", updatedQueue));
    }

    // Skip Token
    @PostMapping("/skip")
    @PreAuthorize("hasAnyRole('admin', 'staff')")
    @Transactional
    public ResponseEntity<?> skipToken(@AuthenticationPrincipal UserPrincipal principal,
                                       @Valid @RequestBody SkipTokenRequest body) {
        UUID bookingId = body.getBookingId();
        UUID slotId = body.getSlotId();

        Optional<Slot> slotOpt = slotRepository.findById(slotId);
        if (slotOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Slot not found"));
        }
        Slot slot = slotOpt.get();

        // Enforce ownership checks
        if ("staff".equals(principal.getRole())) {
            if (principal.getBusinessId() == null || !principal.getBusinessId().equals(slot.getBusinessId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("error", "Access denied: this slot does not belong to your business"));
            }
        } else if ("admin".equals(principal.getRole())) {
            if (slot.getBusiness() == null || !principal.getId().equals(slot.getBusiness().getOwnerId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("error", "Access denied: this slot does not belong to your business"));
            }
        }

        Optional<Booking> bookingOpt = bookingRepository.findById(bookingId);
        if (bookingOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Booking not found"));
        }

        Booking booking = bookingOpt.get();
        if (!booking.getSlotId().equals(slotId)) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("error", "Booking does not belong to the specified slot"));
        }
        booking.setStatus("skipped");
        booking = bookingRepository.save(booking);

        QueueEvent skippedEvent = QueueEvent.builder()
                .bookingId(bookingId)
                .eventType("skipped")
                .build();
        queueEventRepository.save(skippedEvent);

        queueService.invalidateCache(slotId);
        List<Booking> queue = queueService.getQueueState(slotId);

        sseService.broadcastQueue(slot.getBusinessId(), slotId, Map.of("slotId", slotId, "queue", queue));

        return ResponseEntity.ok(Map.of("success", true, "booking", booking, "queue", queue));
    }

    // Complete Service
    @PostMapping("/complete")
    @PreAuthorize("hasAnyRole('admin', 'staff')")
    @Transactional
    public ResponseEntity<?> completeService(@AuthenticationPrincipal UserPrincipal principal,
                                             @Valid @RequestBody CompleteServiceRequest body) {
        UUID bookingId = body.getBookingId();
        UUID slotId = body.getSlotId();

        Optional<Slot> slotOpt = slotRepository.findById(slotId);
        if (slotOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Slot not found"));
        }
        Slot slot = slotOpt.get();

        // Enforce ownership checks
        if ("staff".equals(principal.getRole())) {
            if (principal.getBusinessId() == null || !principal.getBusinessId().equals(slot.getBusinessId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("error", "Access denied: this slot does not belong to your business"));
            }
        } else if ("admin".equals(principal.getRole())) {
            if (slot.getBusiness() == null || !principal.getId().equals(slot.getBusiness().getOwnerId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("error", "Access denied: this slot does not belong to your business"));
            }
        }

        Optional<Booking> bookingOpt = bookingRepository.findById(bookingId);
        if (bookingOpt.isEmpty() || !"serving".equals(bookingOpt.get().getStatus())) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("error", "Booking not found or not currently serving"));
        }

        Booking booking = bookingOpt.get();
        if (!booking.getSlotId().equals(slotId)) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("error", "Booking does not belong to the specified slot"));
        }
        booking.setStatus("done");
        booking = bookingRepository.save(booking);

        QueueEvent doneEvent = QueueEvent.builder()
                .bookingId(bookingId)
                .eventType("done")
                .build();
        queueEventRepository.save(doneEvent);

        queueService.invalidateCache(slotId);
        List<Booking> queue = queueService.getQueueState(slotId);

        sseService.broadcastQueue(slot.getBusinessId(), slotId, Map.of("slotId", slotId, "queue", queue));

        return ResponseEntity.ok(Map.of("success", true, "booking", booking, "queue", queue));
    }

    // Get Queue Status (REST fallback)
    @GetMapping("/status/{slotId}")
    public ResponseEntity<?> getQueueStatus(@PathVariable UUID slotId) {
        List<Booking> queue = queueService.getQueueState(slotId);

        Optional<Slot> slotOpt = slotRepository.findById(slotId);
        int avgTime = 10;
        if (slotOpt.isPresent() && slotOpt.get().getBusiness() != null) {
            avgTime = slotOpt.get().getBusiness().getAvgServiceTime();
        }

        List<Map<String, Object>> enriched = new ArrayList<>();
        int activePos = 1;
        for (Booking b : queue) {
            if (!"done".equals(b.getStatus())) {
                Map<String, Object> item = new HashMap<>();
                item.put("id", b.getId());
                item.put("token_number", b.getTokenNumber());
                item.put("status", b.getStatus());
                item.put("user_id", b.getUserId());
                item.put("users", b.getUser());
                item.put("position", activePos);
                item.put("estimated_wait", queueService.calcWaitTime(activePos - 1, avgTime));
                enriched.add(item);
                activePos++;
            }
        }

        return ResponseEntity.ok(enriched);
    }
}
